crates/burn-derive/src/module/codegen_enum.rs

```diff
@@ -0,0 +1,192 @@
+                Self::#variant(burn::module::Module::<B>::fork(module, device))
+            }
+        });
+
+        quote! {
+            fn fork(self, device: &B::Device) -> Self {
+                #match_body
+            }
+        }
+    }
+
+    fn gen_map(&self) -> TokenStream {
+        let match_body = self.gen_variants_match_fn(|variant| {
+            quote! {
+                Self::#variant(burn::module::Module::<B>::map(module, mapper))
+            }
+        });
+
+        quote! {
+            fn map<Mapper: burn::module::ModuleMapper<B>>(self, mapper: &mut Mapper) -> Self {
+                #match_body
+            }
+        }
+    }
+
+    fn gen_valid(&self) -> TokenStream {
+        let match_body = self.gen_variants_match_fn(|variant| {
+            quote! {
+                Self::InnerModule::#variant(burn::module::AutodiffModule::<B>::valid(module))
+            }
+        });
+
+        quote! {
+            fn valid(&self) -> Self::InnerModule {
+                #match_body
+            }
+        }
+    }
+
+    fn gen_into_record(&self) -> TokenStream {
+        let match_body = self.gen_variants_match_fn(|variant| {
+            quote! {
+                Self::Record::#variant(burn::module::Module::<B>::into_record(module))
+            }
+        });
+
+        quote! {
+            fn into_record(self) -> Self::Record {
+                #match_body
+            }
+        }
+    }
+
+    fn gen_load_record(&self) -> TokenStream {
+        let match_body = self.gen_variants_match_fn(|variant| {
+            quote! {
+                {
+                    let Self::Record::#variant(r) = record else {panic!("Can't parse record from a different variant");};
+                    Self::#variant(burn::module::Module::<B>::load_record(module, r))
+                }
+            }
+        });
+
+        quote! {
+            fn load_record(self, record: Self::Record) -> Self {
+                #match_body
+            }
+        }
+    }
+
+    fn gen_clone(&self) -> TokenStream {
+        let match_body = self.gen_variants_match_fn(|variant| {
+            quote! {
+                Self::#variant(module.clone())
+            }
+        });
+
+        quote! {
+            fn clone(&self) -> Self {
+                #match_body
+            }
+        }
+    }
+
+    fn record_codegen(self) -> Self::RecordCodegen {
+        EnumModuleRecordCodegen::new(self.variants)
+    }
+}
+
+impl EnumModuleCodegen {
+    pub fn from_ast(ast: &syn::DeriveInput) -> Self {
+        Self {
+            variants: parse_variants(ast),
+        }
+    }
+
+    /// Generate the enum variants' match arm with the provided function
+    fn gen_variants_match_fn<F>(&self, func: F) -> TokenStream
+    where
+        F: Fn(Ident) -> TokenStream,
+    {
+        let mut match_arms = quote! {};
+
+        for variant in self.variants.iter() {
+            let name = &variant.ident;
+            let arm_pattern = quote! {Self::#name(module)};
```

Comment:
Yeah only unnamed enum support in this PR. We could definitely add an issue and improve the support, I don't think it would require that much more work now that the table is set.